models/refresh_token.go

```diff
@@ -47,22 +49,41 @@ func GrantRefreshTokenSwap(tx *storage.Connection, user *User, token *RefreshTok
 		if terr = tx.UpdateOnly(token, "revoked"); terr != nil {
 			return terr
 		}
-		newToken, terr = createRefreshToken(rtx, user)
+		newToken, terr = createRefreshToken(rtx, user, token)
 		return terr
 	})
 	return newToken, err
 }
 
+// RevokeTokenFamily revokes all refresh tokens that descended from the provided token.
+func RevokeTokenFamily(tx *storage.Connection, token *RefreshToken) error {
+	err := tx.RawQuery(`
+	with recursive token_family as (
+		select id, user_id, token, revoked, parent from refresh_tokens where parent = ?
```

Comment:
I think it might be worth adding an index on `parent` since the `auth.refresh_tokens` table is typically very large, and filtering by `parent` without an index is going to be very expensive.